assembly/runtime.ts

```diff
@@ -801,26 +842,37 @@ export class RootContext extends BaseContext {
    * @param timeout_milliseconds Timeout for the request, in milliseconds.
    * @param cb Callback to be invoked when the request is complete.
    */
-  httpCall(cluster: string, headers: Headers, body: ArrayBuffer, trailers: Headers,
-    timeout_milliseconds: u32, cb: HttpCallback): WasmResultValues {
-
+  httpCall(cluster: string, headers: Headers, body: ArrayBuffer, trailers: Headers, timeout_milliseconds: u32, origin_context_id: u32, cb: (origin_context_id:u32, headers: u32,  body_size: usize, trailers: u32) => void): WasmResultValues {
     let buffer = String.UTF8.encode(cluster);
     let header_pairs = serializeHeaders(headers);
     let trailer_pairs = serializeHeaders(trailers);
-    let token = globalU32Ref;
+    let token = new Reference<u32>();
+    this.http_calls_.set(token.data, new HttpCallback(origin_context_id, cb));
+    log(LogLevelValues.debug, "http_calls_: " + this.http_calls_.size.toString());
+    log(LogLevelValues.debug, "set token: " + token.data.toString() + " on " + this.context_id.toString());
     let result = imports.proxy_http_call(changetype<usize>(buffer), buffer.byteLength, changetype<usize>(header_pairs), header_pairs.byteLength, changetype<usize>(body), body.byteLength, changetype<usize>(trailer_pairs), trailer_pairs.byteLength, timeout_milliseconds, token.ptr());
-    if (result == WasmResultValues.Ok) {
-      this.http_calls_.set(token.data, cb);
-    }
+    log(LogLevelValues.debug, "result: " + result.toString());
     return result;
   }
+  
   onHttpCallResponse(token: u32, headers: u32, body_size: u32, trailers: u32): void {
-    if (this.http_calls_.has(token)) {
-      let callback = this.http_calls_.get(token);
-      this.http_calls_.delete(token);
-      callback.call();
+    log(LogLevelValues.debug, "context id: " + this.context_id.toString() + ": onHttpCallResponse(token: " + token.toString() + ", headers:" + headers.toString() + ", body_size:" + body_size.toString() + ", trailers:" + trailers.toString() + ")");
+    log(LogLevelValues.debug, "http_calls_: " + this.http_calls_.size.toString());
+    log(LogLevelValues.debug, "get token: " + token.toString() + " from " + this.context_id.toString());
+    log(LogLevelValues.debug, "context_map: " + context_map.keys().join(", "));
+    const readToken = get_buffer_bytes(BufferTypeValues.HttpCallResponseBody, 0, token);
+    log(LogLevelValues.debug, "readToken: " + readToken.toString());
+    if (this.http_calls_.has(0)) {
+      let callback = this.http_calls_.get(0);
+      log(LogLevelValues.debug, "onHttpCallResponse: calling callback for context id: " + callback.origin_context_id.toString());
+      this.http_calls_.delete(0);
+      this.setEffectiveContext(callback.origin_context_id);
```

Comment:
i'm not sure if we want to set the effective context id here, as i dont know if you will have access to the response headers/body from the http call once you switch the context.
see this c++ example for reference:
https://github.com/envoyproxy/envoy-wasm/blob/dfb104874b3babc7e676aaa0199f2926410c0671/test/extensions/filters/http/wasm/test_data/async_call_cpp.cc#L15

Reply:
Yeah, I saw that comment when I was investigating this, but I think we can read the headers/body/trailers content of the response first, save it in memory, change context, and then call the callback function with the variables we already saved.

I like it this way because the Filter is not aware of this context manipulation. It makes it cleaner on the user of the SDK side.

Reply:
yes, but this means that each filter will "pay" the cost of copying the data, which might hurt performance

Reply:
Yeahhhh. But I still don't think we should allow the Filters to call the proxy directly. I was trying to avoid that at all costs. 
I'm not so sure about this now. I'm divided...

Reply:
i agree with not wanting the filter to call the `proxy_` api directly. 
note that there is a wrapper for `proxy_get_buffer_bytes` (see here: https://github.com/solo-io/proxy-runtime/blob/4db3a19f41293387c92881c7481197796421eb44/assembly/runtime.ts#L540)
so the filter can use this as convienence.
to minimize error we can maybe add specific ones? i.e. getHttpCallBuffer? not sure if that would make sense or burden the user with more things to learn?

Reply:
I like that. That should work...
But we need to work on standardizing how data is read from the buffers. I, myself, need to learn better how you are doing that for the request headers. That work needs to be extended and reused everywhere data is exchanged between the proxy and the Filter.

Reply:
sounds good; i do think that in the mean time we should not call setEffectiveContext id from here, and leave it to the callback